### PR TITLE
Switch to non-recursive on heap virtual stack when building logical plan from SQL expression

### LINKED ISSUE #1434: Query with 100 OR conditions overflows stack
**Describe the bug**
`SELECT * FROM table WHERE <condition0> OR <condition1> OR ...` succeeds with 50 conditions but overflows stack with 100 conditions.

**To Reproduce**
Adding this test to arrow-datafusion/datafusion/tests/sql.rs demonstrates the issue:

```
#[tokio::test]
async fn query_with_many_conditions() -> Result<()> {
    let schema = Arc::new(Schema::new(vec![Field::new("c1", DataType::Utf8, true)]));
    let data = RecordBatch::try_new(
        schema.clone(),
        vec![Arc::new(StringArray::from(vec!["foo", "bar"]))],
    )
    .unwrap();

    let table = MemTable::try_new(schema, vec![vec![data]])?;

    let mut ctx = ExecutionContext::new();
    ctx.register_table("test", Arc::new(table)).unwrap();

    let num_conditions = 50;
    println!("attempted a query with {:?} conditions...", num_conditions);
    let where_clause = (0..num_conditions)
        .map(|i| format!("c1 = 'value{:?}'", i))
        .collect::<Vec<String>>()
        .join(" OR ");
    let sql = format!("SELECT * from test where {};", where_clause);
    execute_to_batches(&mut ctx, &sql).await;
    println!("query succeeded with {:?} conditions", num_conditions);

    let num_conditions = 100;
    println!("attempted a query with {:?} conditions...", num_conditions);
    let where_clause = (0..num_conditions)
        .map(|i| format!("c1 = 'value{:?}'", i))
        .collect::<Vec<String>>()
        .join(" OR ");
    let sql = format!("SELECT * from test where {};", where_clause);
    execute_to_batches(&mut ctx, &sql).await;
    println!("query succeeded with {:?} conditions", num_conditions);
    Ok(())
}
```

Output:
```
running 1 test
attempted a query with 50 conditions...
query succeeded with 50 conditions
attempted a query with 100 conditions...

thread 'query_with_many_conditions' has overflowed its stack
fatal runtime error: stack overflow
error: test failed, to rerun pass '--test sql'
```

**Expected behavior**
Expected query to succeed.

### LINKED ISSUE #1444: Avoid stack overflow in Datafusion
Rewrite recursive procedures into iterative procedures, suggested by @houqp @alamb 

> > This is an acceptable short term workaround, but I think it would be more efficient and more elegant if we rewrite these recursive procedures into iterative procedures.

> Do you mean to use push-based model？

@xudong963 , I think what @houqp  is suggesting is to rewrite code that is recursive to not be recursive.

The pattern for Datafusion probably looks like taking code like

```rust
fn visit(expr: &expr)  {
  for child in expr.children() {
    visit(child)
  }
  // do actual expr logic
}
```

And changing it so the state is tracked with a structure on the heap rather than a stack. I think [`VecDeque`](https://doc.rust-lang.org/stable/std/collections/struct.VecDeque.html) is a good one for rust:

```rust
fn visit(expr: &expr)  {
  let mut worklist = VecDequeue::new();
  worklist.push_back(expr);
  while !worklist.is_empty() {
    let parent = worklist.pop_front();
    for child in parent.children() {
      worklist.push_back(child)
    }
    // do actual expr logic on parent
  }
}
```

(aka avoid the call back to `visit`)

_Originally posted by @alamb in https://github.com/apache/arrow-datafusion/issues/1437#issuecomment-992770032_

### LINKED ISSUE #6040: Stack overflow while generating logical plan from statement
### Describe the bug

While using multiple conditions are used, a stack overflow error occurs.

In particular, when used with tokio, more limitations arise because the [default stack size is 2MiB](https://asomers.github.io/tokio-file/tokio/runtime/struct.Builder.html#method.thread_stack_size).

### To Reproduce

I referred to reproduce code from issue #1434 provided by @mcassels.
`SELECT * FROM table WHERE <condition0> OR <condition1> OR ...` 
``` rust
use datafusion::{
    arrow::datatypes::{DataType, Field, Schema},
    common::Result,
    config::ConfigOptions,
    error::DataFusionError,
    logical_expr::{
        logical_plan::builder::LogicalTableSource, AggregateUDF, ScalarUDF, TableSource,
    },
    sql::{
        planner::{ContextProvider, SqlToRel},
        sqlparser::{dialect::GenericDialect, parser::Parser},
        TableReference,
    },
};
use std::{collections::HashMap, sync::Arc};

#[global_allocator]
static GLOBAL: mimalloc::MiMalloc = mimalloc::MiMalloc;

#[tokio::main]
async fn main() -> Result<()> {
    let num_conditions = 255;
    let where_clause = (0..num_conditions)
        .map(|i| format!("column1 = 'value{:?}'", i))
        .collect::<Vec<String>>()
        .join(" OR ");
    let sql = format!("SELECT * from table1 where {};", where_clause);
    get_optimized_plan(sql).await?;
    println!("query succeeded with {:?} conditions", num_conditions);

    let num_conditions = 256;
    let where_clause = (0..num_conditions)
        .map(|i| format!("column1 = 'value{:?}'", i))
        .collect::<Vec<String>>()
        .join(" OR ");
    let sql = format!("SELECT * from table1 where {};", where_clause);
    get_optimized_plan(sql).await?;
    println!("query succeeded with {:?} conditions", num_conditions);

    Ok(())
}

async fn get_optimized_plan(sql: String) -> Result<()> {
    let schema_provider = TestSchemaProvider::new();

    let dialect = GenericDialect {};
    let ast = Parser::parse_sql(&dialect, &sql).unwrap();
    let statement = &ast[0];
    let sql_to_rel = SqlToRel::new(&schema_provider);
    sql_to_rel.sql_statement_to_plan(statement.clone()).unwrap();

    Ok(())
}

struct TestSchemaProvider {
    options: ConfigOptions,
    tables: HashMap<String, Arc<dyn TableSource>>,
}

impl TestSchemaProvider {
    pub fn new() -> Self {
        let mut tables = HashMap::new();
        tables.insert(
            "table1".to_string(),
            create_table_source(vec![Field::new(
                "column".to_string(),
                DataType::Utf8,
                false,
            )]),
        );

        Self {
            options: Default::default(),
            tables,
        }
    }
}

fn create_table_source(fields: Vec<Field>) -> Arc<dyn TableSource> {
    Arc::new(LogicalTableSource::new(Arc::new(
        Schema::new_with_metadata(fields, HashMap::new()),
    )))
}

impl ContextProvider for TestSchemaProvider {
    fn get_table_provider(&self, name: TableReference) -> Result<Arc<dyn TableSource>> {
        match self.tables.get(name.table()) {
            Some(table) => Ok(table.clone()),
            _ => Err(DataFusionError::Plan(format!(
                "Table not found: {}",
                name.table()
            ))),
        }
    }

    fn get_function_meta(&self, _name: &str) -> Option<Arc<ScalarUDF>> {
        None
    }

    fn get_aggregate_meta(&self, _name: &str) -> Option<Arc<AggregateUDF>> {
        None
    }

    fn get_variable_type(&self, _variable_names: &[String]) -> Option<DataType> {
        None
    }

    fn options(&self) -> &ConfigOptions {
        &self.options
    }
}
```
Output
``` bash
query succeeded with 255 conditions

thread 'main' has overflowed its stack
fatal runtime error: stack overflow
```

If there are more than 256 conditions, stack overflow occurs. This happens only `debug mode`, related to https://github.com/apache/arrow-datafusion/issues/1434#issuecomment-992758421. 

### Expected behavior

Work without overflows..

### Additional context

I guess 2 approaches to this problem.

Approach#1
Parameters are received as reference or without using box pointers in some functions, such as select_to_plan and plan_selection.  This maybe can make Stack grow faster.

And I found some stack allocation with enumeration.
https://www.reddit.com/r/rust/comments/zbla3j/how_does_enums_work_where_are_they_allocated/

Approach#2
Using [Address Sanitizer](https://doc.rust-lang.org/beta/unstable-book/compiler-flags/sanitizer.html) with the above example, error occurred in fmt::Display. But, I'm not sure exactly where it happened.

This would be related to rust issue: https://github.com/rust-lang/rust/issues/45838 .

### PR DESCRIPTION
**Pull request by aprimadi**
# Which issue does this PR close?

Closes #6040
Closes #1444 
Closes #1434 

# Rationale for this change

Not sure why anyone would want more than 256 expressions but I thought I would write the iterative version just for fun. Now users can write all the expressions they want (or at least until the generated `Expr` can no longer fit in the stack).

Kudos to @alamb and @houqp for suggesting this iterative version.

# What changes are included in this PR?

- Iterative version of expr logical plan builder
- Add tests verifying the number of expressions that can be handled is now larger

# Are these changes tested?

Yes. Tested by existing tests. Also added a few stack overflow tests.

# Are there any user-facing changes?

No

### CODE REVIEW

File: datafusion/sql/src/expr/mod.rs [hunk 1]
```diff
@@ -46,27 +46,56 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
-            SQLExpr::BinaryOp { left, op, right } => {
-                self.parse_sql_binary_op(*left, op, *right, schema, planner_context)
+        enum StackEntry {
+            SQLExpr(Box<SQLExpr>),
+            Operator(Operator),
+        }
+
+        // Virtual stack machine to convert SQLExpr to Expr
+        // This allows visiting the expr tree in a depth-first manner which
+        // produces expressions in postfix notations, i.e. `a + b` => `a b +`.
```
Comment (alamb) [inline datafusion/sql/src/expr/mod.rs:56]: Can you please add a link to https://github.com/apache/arrow-datafusion/issues/1444 (so the rationale for using this stack machine is clearer):

```suggestion
        // produces expressions in postfix notations, i.e. `a + b` => `a b +`.
        // See https://github.com/apache/arrow-datafusion/issues/1444
```

File: datafusion/sql/src/expr/mod.rs [hunk 2]
```diff
@@ -46,27 +46,56 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
+        let mut stack: Box<Vec<StackEntry>> =
+            Box::new(vec![StackEntry::SQLExpr(Box::new(sql))]);
+        let mut eval_stack = Box::<Vec<Expr>>::default();
```
Comment (alamb) [inline datafusion/sql/src/expr/mod.rs:59]: I don't think `Box`ing the `Vec` is necessary in this case (as there is a `Box` internally in the `Vec` for the actual data items).

I tried switching it to 
```suggestion
        let mut stack = vec![StackEntry::SQLExpr(Box::new(sql))];
        let mut eval_stack = vec![];
```

And all the tests continue to pass

File: datafusion/sql/src/expr/mod.rs [hunk 3]
```diff
@@ -566,3 +595,124 @@ fn plan_indexed(expr: Expr, mut keys: Vec<SQLExpr>) -> Result<Expr> {
...
+                        sql_expr,
+                        &schema,
+                        &mut planner_context,
+                    ).unwrap();
+                }
+            }
+        };
+    }
+
+    test_stack_overflow!(64);
```
Comment (alamb) [inline datafusion/sql/src/expr/mod.rs:710]: I verified that without the code in this PR, the tests fail as expected:

```
running 14 tests
test expr::arrow_cast::test::test_parse_data_type_whitespace_tolerance ... ok(B
test expr::identifier::test::test_form_identifier ... ok(B
test expr::identifier::test::test_generate_schema_search_terms ... ok(B
test expr::arrow_cast::test::parse_data_type_errors ... ok(B
test expr::arrow_cast::test::test_parse_data_type ... ok(B
test parser::tests::create_external_table ... ok(B
test expr::tests::test_stack_overflow_64 ... ok(B
test expr::tests::test_stack_overflow_128 ... ok(B
test expr::tests::test_stack_overflow_256 ... ok(B
test expr::tests::test_stack_overflow_512 ... ok(B

thread 'expr::tests::test_stack_overflow_1024' has overflowed its stack
fatal runtime error: stack overflow
error: test failed, to rerun pass `-p datafusion-sql --lib`

Caused by:
  process didn't exit successfully: `/Users/alamb/Software/target-df2/debug/deps/datafusion_sql-8a77caa8d397a04b` (signal: 6, SIGABRT: process abort signal)

```

**Review by alamb — APPROVED** [review]
Thank you very much @aprimadi . This is really neat. 

I guess the next level would be to thread the stack machine down to parsing all expressions (not just binary ops) which while would make the code more uniform I am not sure it would make it easier to understand

> Not sure why anyone would want more than 256 expressions

LOL I think this kind of expression is generated by other programs / tools like BI frontends

File: datafusion/sql/src/expr/mod.rs (hunk 2 above)
Comment (aprimadi) [reply datafusion/sql/src/expr/mod.rs:60]: Nice. I didn't know that.

**Comment by alamb** [comment]
Thanks again @aprimadi

**Comment by aprimadi** [comment]
Thank you @alamb for the review

**Pull request merged**